File: solutions/cpp/0632.cpp
struct Item {
  int i;
  int j;
};

class Solution {
 public:
  vector<int> smallestRange(vector<vector<int>>& nums) {
    auto compare = [&](const Item& a, const Item& b) {
      return nums[a.i][a.j] > nums[b.i][b.j];
    };
    priority_queue<Item, vector<Item>, decltype(compare)> pq(compare);

    int maxi = INT_MIN;
    for (int i = 0; i < nums.size(); ++i) {
      pq.push({i, 0});
      maxi = max(maxi, nums[i][0]);
    }
    int mini = nums[pq.top().i][pq.top().j];
    int ansMin = mini;
    int ansMax = maxi;

    while (pq.size() == nums.size()) {
      const auto [i, j] = pq.top(); pq.pop();
      if (j + 1 < nums[i].size()) {
        pq.push({i, j + 1});
        maxi = max(maxi, nums[i][j + 1]);
        mini = nums[pq.top().i][pq.top().j];
        if (maxi - mini < ansMax - ansMin) {
          ansMin = mini;
          ansMax = maxi;
        }
      }
    }

    return {ansMin, ansMax};
  }
};

File: solutions/java/0632.java
class Solution {
  public int[] smallestRange(List<List<Integer>> nums) {
    Queue<Elem> pq = new PriorityQueue<>((e1, e2) -> e1.value - e2.value);

    int maxRange = Integer.MIN_VALUE;
    for (int i = 0; i < nums.size(); ++i) {
      pq.add(new Elem(nums.get(i).get(0), i, 0));
      maxRange = Math.max(maxRange, nums.get(i).get(0));
    }
    int minRange = pq.peek().value;

    int[] ans = new int[] { minRange, maxRange };

    while (pq.size() == nums.size()) {
      Elem elem = pq.poll();
      if (elem.c + 1 < nums.get(elem.r).size()) {
        pq.add(new Elem(nums.get(elem.r).get(elem.c + 1), elem.r, elem.c + 1));
        maxRange = Math.max(maxRange, nums.get(elem.r).get(elem.c + 1));
        minRange = pq.peek().value;
        if (maxRange - minRange < ans[1] - ans[0]) {
          ans[0] = minRange;
          ans[1] = maxRange;
        }
      }
    }

    return ans;
  }

  private class Elem {
    int value;
    int r;
    int c;

    public Elem(int value, int r, int c) {
      this.value = value;
      this.r = r;
      this.c = c;
    }
  }
}